# Numbers


Operations on numbers

<table style="text-align : left ">
  <tr>
    <th>Operations</th>
    <th>Explanation</th>
  </tr>
  <tr>
    <td>-</td>
    <td>Minus two numbers</td>
  </tr>
  <tr>
    <td>+</td>
    <td>Add two numbers</td>
  </tr>
    <tr>
    <td>/</td>
    <td>Divides two numbers</td>
  </tr>
    <tr>
    <td>*</td>
    <td>Multiplies two numbers</td>
  </tr>
    <tr>
    <td>//</td>
    <td>Floor division</td>
  </tr>
    <tr>
    <td>**</td>
    <td>Exponent</td>
  </tr>
   <tr>
    <td>%</td>
    <td>Divides numbers(returns remainder)</td>
  </tr>
</table>

In [ ]:
# Addtion
1 + 1

2

In [ ]:
# Subtraction 
6 - 0.5

5.5

In [ ]:
# Division vs Floor Division
print(11 / 2) # Division
print(11 // 2) # Floor Division

5.5
5


In [ ]:
# Multiplication
5 * 5

25

In [ ]:
# Exponent
5 ** 5

3125

In [ ]:
# Modulus
print(19 % 10)

9


In [ ]:
(1 + 1) * 2

4

In [ ]:
1 + 1 * 2

3

# Floats

<b>A float is basically the same thing as a Number except it has a decimal point in it</b>

In [ ]:
5 / 2 # Returns Float

2.5

5.2<br>
6.3<br>
10.2<br>
7.5<br>
22.2<br>
13202.223<br>
2323.2232 all are floats

# Strings

<b>Strings are a series of characters wrapped in quotes they may be single double or even triple. They are immutable.</b>

In [ ]:
My name is Jagnath.# Shall not work

SyntaxError: invalid syntax (<ipython-input-39-57ba8a524f8e>, line 1)

In [ ]:
print('My name is Aarav Sharma') # Single Quotes

My name is Aarav Sharma


In [ ]:
print("My favourite book is 'The dead'. ")
# Double Quotes needed else python is confused.

My favourite book is 'The dead!!'


In [ ]:
print('''He said that, "I like 'The dead' Book." ''') # Needs triple quotes

He said that, "I like 'The dead' Book." 


<b>A quick note on Variables before we continue strings</b>

# Variables

<b>Variables are labels that store data<br>
eg. a = 12
</b>

In [ ]:
# Interest Calculation
amount = 1000
interest = 0.2
amount_needed_to_pay = amount * interest
print(amount_needed_to_pay)

200.0


Rules of making variables-<br>
•	 It can be only one word with no spaces.<br>
•	 It can use only letters, numbers, and the underscore (_) character.<br>
•	 It can’t begin with a number.

<b>String Indexing and Slicing
<b>

In [ ]:
a = 'My name is Aarav Sharma'

In [ ]:
# Goal Take 'Aarav Sharma' Note: String indexing starts with 0
a[11:]

'Aarav Sharma'

In [ ]:
# Reverse Indexing
a[-12:]

'Aarav Sharma'

In [ ]:
# Goal take alternate charcters from 'Aarav Sharma'
a[11::2]

'ArvSam'

In [ ]:
# Goal: Reverse String
a[::-1]

'amrahS varaA si eman yM'

In [ ]:
# Goal: Reverse the string returned by the reversing the string.
a[::-1][::-1]

'My name is Aarav Sharma'

<b>Common methods on String object-
<br>
In Python, a method is a function that is available for a given object because of the object's type.</b>

In [ ]:
# Create String
name = 'Aarav Sharma'

In [ ]:
# upper() returns a string where all characters are in upper case.
name.upper() 

'AARAV SHARMA'

In [ ]:
# Does not change original string
name

'Aarav Sharma'

In [ ]:
name.lower() #Lowers all characters in string 

'aarav sharma'

In [ ]:
print(name.isupper()) # Check if all letters are Uppercase
print(name.islower()) # Checks if all letters are lowercase

False
False


In [ ]:
print(name.startswith('Aa')) 
print(name.endswith('rma'))

True
True


In [ ]:
name.zfill(20) # Fills in zeros as starting characters till desired length of string not reached

'00000000Aarav Sharma'

In [ ]:
name.split('a') # Splits on letter 'a'

['A', 'r', 'v Sh', 'rm', '']

Tip: after writing 'name.' click tab. Click Shift + Tab for DocString

<b>Concatination of strings</b>

In [ ]:
my_name = 'Aarav Sharma'
prefix = 'My name is ' # Notice the extra space after 'is'
prefix + my_name 

'My name is Aarav Sharma'

In [ ]:
prefix * 10 

'My name is My name is My name is My name is My name is My name is My name is My name is My name is My name is '

# String Interpelation

<b> It is a method used to join strings with other strings </b>

<br><b>With the format method</b>

In [ ]:
print('His name is {}. He is {} years old'.format('Jagnath', 12))

His name is Jagnath. He is 12 years old


<b>Can assign with index and name also-</b>

In [ ]:
print('His name is {}. He is {} years old'.format(12,'Jagnath' ))# WRONG

His name is 12. He is Jagnath years old


In [ ]:
# Right
print('His name is {1}. He is {0} years old'.format(12,'Jagnath' ))
# or 
print('His name is {name}. He is {age} years old'.format(age = 12,name = 'Jagnath' ))

His name is Jagnath. He is 12 years old
His name is Jagnath. He is 12 years old


<b>f'' string method</b>

In [ ]:
print('His name is {}. He is {} years old.'.format('Jagnath', 12))
# Can also be done by-
name = 'Jagnath'
age = 12
print(f'His name is {name}. He is {age} years old')

His name is Jagnath. He is 12 years old

His name is Jagnath. He is 12 years old


<b>%s method</b>

In [ ]:
name = 'Jagnath'
age = 12
prefix = 'His name is %s. His age is %s' # %S AWAITS A FILLER
print(prefix % (name,age))

His name is Jagnath. His age is 12


<b>Concatination</b>

In [ ]:
name = 'Jagnath'
prefix = 'His name is '
print(prefix + name)

His name is Jagnath


<b>Some Useful functions on strings</b>

In [ ]:
name = 'Jagnath'
print(len(name)) # Returns length of string
print(name) # prints name
input('Type what is 1+1') # Awaits user input

7
Jagnath
Type what is 1+12


'2'

<b>Immutability of string</b>
<br>

Immutability means non-changable. A string cannot be changed directly.

In [ ]:
name = 'Jagnath'
# Goal: Make Jagnath - Sagnath
name[0] = 'S' # Wont work 'str' object does not support item assignment

TypeError: 'str' object does not support item assignment

Later on, we will see many datatypes that support such a item assignment

# Lists

Lists are ordered sequences used to store multiple items in a single variable. They are Mutable.

In [ ]:
# Defining a list
my_list = ['Blueberry', 123, 1.223, [1, 2, 3, 4, 5, 6, 7, 8]] 
# Can store any DataType (Even another lists)

<b>List indexing and slicing</b>

Just like strings we can also index and slice lists

In [ ]:
my_list = ['Dog', 'Cat', 2 , 7.0, 99, 'The Dog', 'The human', 'Every...']

In [ ]:
# Goal: Grab 'Cat' Note lists' start inex is 0
my_list[1]

'Cat'

In [ ]:
# Goal Grab:- 'The human'
my_list[6]

'The human'

In [ ]:
# Goal: Grab 'Cat', 2 ,7.0, 99
my_list[1:7] # Excludes last index

['Cat', 2, 7.0, 99, 'The Dog', 'The human']

In [ ]:
# Reverse Indexing
my_list[1:-1]

['Cat', 2, 7.0, 99, 'The Dog', 'The human']

In [ ]:
# Goal: Grab every alternate item
my_list[::2]

['Dog', 2, 99, 'The human']

In [ ]:
# Goal Reverse list
my_list[::-1]

['Every...', 'The human', 'The Dog', 99, 7.0, 2, 'Cat', 'Dog']

<b>List methods</b>

In [ ]:
# Define lists
a = [1, 2, 3, 4, 5, 6]
b = [7, 8, 9, 10]

In [ ]:
# Concatenate lists
a + b
# Does not Change a or b

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [ ]:
# a and b not changed
print(a)
print(b)

[1, 2, 3, 4, 5, 6]
[7, 8, 9, 10]


In [ ]:
a * 2 # Multiply lists

[1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6]

In [ ]:
a = [1, 2, 3, 4, 5, 6]
a.append(10)
a
# adds number 10 to list HAPPENS IN PLACE

In [ ]:
a = [1, 2, 3, 4, 5, 6]
a.extend([10])
a
# adds number 10 to list HAPPENS IN PLACE

[1, 2, 3, 4, 5, 6, 10]

<b>.extend() vs .append(). They appear same but have one difference:- .extend acts can add two lists as + operator while append acts as adding something</b>

In [ ]:
# .extend() vs .append()
a = [1, 2, 3, 4, 5, 6, 7]
b = [8, 9, 10]
a.append(b)
print(a)

[1, 2, 3, 4, 5, 6, 7, [8, 9, 10]]


In [ ]:
a = [1, 2, 3, 4, 5, 6, 7]
b = [8, 9, 10]
a.extend(b)
print(a)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [ ]:
a = [1, 2, 3, 4, 5, 6]
a.remove(1) # Removes item from list HAPPENS IN PLACE
print(a)

[2, 3, 4, 5, 6]


<b>A quick note on what do I mean by 'Happen In Place'</b>

# Methods that return or make inplace changes and NoneType

There are some methods that make inplace changes such as the .append() method. These methods make cnages inplace i.e. they change the object directly and when we print the object we see it being changed.

In [ ]:
# Such as-
a = [1,2,3,4,5,6]
a.append(10)
print(a) # Or just type 'a' as in Jupyter Notebook System

[1, 2, 3, 4, 5, 6, 10]


This makes direct changes it does not give anything.

In [ ]:
a = [1,2,3,4,5,6]
print(a.append(10))
a

None


[1, 2, 3, 4, 5, 6, 10]

None is returned as this method doesn't give anything.
None is used to define a null value. In python 'a function that does not return something' when called by print, prints None.

Let's see a method that does return something.

In [ ]:
a = 'My name is Jagnath.'
print(a.startswith('My'))

True


This method doesn't change anything so, it returns true or false.

<b> .sort() method can sort strings as well as numbers</b>

In [ ]:
# Sort numbers
a = [9, 4, 2, 6, 3, 6]
a.sort()
a

[2, 3, 4, 6, 6, 9]

In [ ]:
# Sort strings on alphabetical order
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
a.sort()
a

['Cat', 'Dog', 'Every...', 'The Dog', 'The human']

In [ ]:
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
a.reverse()
#Same as [::-1] except that it HAPPENS IN PLACE
a

['Every...', 'The human', 'The Dog', 'Cat', 'Dog']

In [ ]:
# Removes Object from list and returns it (By Default deltes and returnes last item)
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
var = a.pop()
print(var)

In [ ]:
# Can also specipy which item to .pop()
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
var = a.pop(2)# List Index Starts with 0
print(var)

The Dog


In [ ]:
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
a.clear() # deletes all items
a

[]

In [ ]:
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
a.index('Cat') # returns index of passed obj in list

1

In [ ]:
a = [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3]
a.count(0) # Counts number of items in list

9

In [ ]:
a.copy()# creates copy of list

[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3]

In [ ]:
a = my_list = ['Dog', 'Cat', 'The Dog', 'The human', 'Every...']
a.insert(1, 'His name is') # Inserts obj at specified index. First arg is index second is obj
#Unlike .append() that adds to last.
print(a)

['Dog', 'His name is', 'Cat', 'The Dog', 'The human', 'Every...']


# str(), int(), and float() Functions

<b>They are used to convert datatypes we learnt about into other DataTypes</b>

Conversion of digits and floats to strings

In [ ]:
str(11) # Integer

'11'

In [ ]:
str(11.11)

'11.11'

Conversion of strings and floats to Integers

In [ ]:
int('11')

11

In [ ]:
int(11.23)

11

In [ ]:
int(11.95)

11

In [ ]:
int('11.11') # Cannot convert a string which has a floating num inside.
# need to do - int(float('11.11')) to work

ValueError: invalid literal for int() with base 10: '11.11'

Conversion of strings and integers to floats

In [ ]:
float(11)

11.0

In [ ]:
float(12.12)

12.12

In [ ]:
float('11.11')

11.11

In [ ]:
float('11')

11.0

<b>This Program shows why knowledge of conversion of DataTypes into other DataTypes is useful.</b>

In [ ]:
# Program to add two numbers
a = input('First number: ')
b = input('Second number: ')
print(a+b)

First number: 10
Second number: 10
1010


Why did this happen?
This happened as input() func takes user input. The user input is automatically converted to a string so essentially '10' + '10' = '1010'
We can use the type() func to prove this-

In [ ]:
a = input('First number: ')
b = input('Second number: ')
print(type(a))
print(type(b))

First number: 10
Second number: 10
<class 'str'>
<class 'str'>


So python seeing they are string concatenates them together.
So the correct and useful code would be to convert it into integer and then add them.

In [ ]:
# Correct Code
# Proogram to add two numbers
a = int(input('First number: '))
b = int(input('Second number: '))
print(a+b)

First number: 10
Second number: 10
20


Wouldn't it be nice if our program could also accept floats? Well we could use float which is more flexible.

In [ ]:
a = float(input('First number: '))
b = float(input('Second number: '))
print(a+b)

First number: 10.2
Second number: 6.5
16.7


# Booleans and Comparison Operators

<b>Booleans are used to represent the truth value of an expression.
Baically they are True and False (Notice the Capitalization)</b>

In [ ]:
type(False) 

bool

<b>Comparison Operators</b>

<table style="text-align : left ">
  <tr>
    <th>Operator</th>
    <th>Explanation</th>
  </tr>
  <tr>
    <td>==</td>
    <td>Checks for equality</td>
  </tr>
  <tr>
    <td>></td>
    <td>Checks if a number is greater than another</td>
  </tr>
    <tr>
    <td><</td>
    <td>Checks if a obj is smaller than another obj</td>
  </tr>
    <tr>
    <td>>=</td>
    <td>Checks if a obj is greater than or equal to another obj</td>
  </tr>
    <tr>
    <td><=</td>
        <td>Checks if a obj is smaller than or equal to another ob</td>
  </tr>
    <tr>
    <td>!=</td>
    <td>Checks if a obj is not equal to another obj</td>
  </tr>
    <tr>
    <td><></td>
    <td>Checks if a obj is not equal to another obj(same as != but is not prefered)</td>
  </tr>
</table>

In [ ]:
12 > 11

True

In [ ]:
'zomine' < 'zz' # Checks by alphabetical order(later the alphabet in the alphabets greater its value)

True

In [ ]:
14 > 15 

False

In [ ]:
11 < 10

False

In [ ]:
11 == 10 # notice we write == (Since, do we mean to assign a variable?)

False

In [ ]:
'NAA' == 'NAA'

True

In [ ]:
11 >= 11 # >= means greater than or equal to. 11 == 11 so eval to True

True

In [ ]:
11 <= 11

True

<b>Logical Opeators</b>

<table style="text-align : left ">
  <tr>
    <th>Operator</th>
    <th>Explanation</th>
  </tr>
  <tr>
    <td>and</td>
    <td>Returns True if both statements evaluate to True else: False</td>
  </tr>
    <tr>
    <td>or</td>
    <td>Returns True if either statements evaluate to True else: False</td>
  </tr>
    <tr>
    <td>not</td>
    <td>Returns True if Statement is not True</td>
  </tr>
</table>

<b>With 'and'</b>

In [ ]:
1 + 1 == 2 and 'ABCD' == 'abcd'

False

In [ ]:
1 + 1 == 2 and 'ABCD' == 'ABCD'

True

<b>With 'or'</b>

In [ ]:
1 + 1 == 2 or 'ABCD' == 'abcd'

True

In [ ]:
11 - 11 == -1 or 1 + 1 == 3

False

<b>With 'not'</b>

In [ ]:
not 11 - 11 == -1 and not 1 + 1 == 3

True

In [ ]:
not 1 + 1 == 2 and 1 == 2

False